src/FsCheck/ReflectArbitrary.fs

```diff
@@ -59,40 +61,49 @@ module internal ReflectArbitrary =
                     else
                         if List.exists(fun (x : Type) -> x.ToString() = t.ToString()) ts then 2 else 1
                         
-                let unionGen create ts =
+                let unionGen create ts getGenerator =
                     let productGen (ts : list<Type>) =
-                        let gs = [ for t in ts -> getGenerator t ]
+                        let gs = ts |> List.map getGenerator
                         let n = gs.Length
-                        [ for g in gs -> sized (fun s -> resize ((s / n) - 1) g )]
+                        [| for g in gs -> sized (fun s -> resize ((s / n) - 1) g )|]
                     let g = productGen ts
                     let res = g |> sequence |> map (List.toArray >> create)
                     res
 
-                let gs = [ for _,(_,fields,create,_) in getUnionCases t -> unionSize fields, lazy (unionGen create fields) ]
-                let lowest = List.reduce min <| List.map fst gs
-                let small() = [ for i,g in gs do if i = lowest then yield g.Force() ]
-                let large() = [ for _,g in gs -> g.Force() ]
-                let getgs size = 
-                    if size <= 0 then small() else large() 
-                    |> oneof 
-                    |> resize (size - 1) 
-                sized getgs |> box
+                let gs = [| for _,(_,fields,create,_) in getUnionCases t -> unionSize fields, unionGen create fields |]
```

Comment:
Most changes were trivial but this was a bit tricky. So here I really like some feedback.